packages/@glimmer/reference/test/iterable-test.ts

```diff
@@ -1,530 +1,186 @@
+  module('iterator delegates', () => {
+    test('it correctly iterates delegates', assert => {
+      let obj = { a: 'Yehuda', b: 'Godfrey' };
+      let target = new IterableWrapper(obj);
 
-  arr = arr.slice();
-  arr.reverse();
-  reference.update(arr);
+      assert.deepEqual(target.toValues(), objectValues(obj));
+    });
 
-  sync(target, artifacts);
+    test('it correctly synchronizes delegates when changed', assert => {
+      let obj = { a: 'Yehuda', b: 'Godfrey' } as any;
+      let target = new IterableWrapper(obj);
 
-  assert.deepEqual(target.toValues(), arr);
+      assert.deepEqual(target.toValues(), objectValues(obj));
 
-  reference.update([
-    { key: 'a', name: 'Tom' },
-    { key: 'b', name: 'Stef ' },
-  ]);
+      obj.c = 'Rob';
 
-  sync(target, artifacts);
+      assert.deepEqual(target.toValues(), objectValues(obj));
 
-  assert.deepEqual(target.toValues(), [
-    { key: 'a', name: 'Tom' },
-    { key: 'b', name: 'Stef ' },
-  ]);
+      obj.a = 'Godhuda';
 
-  arr = arr.slice();
-  arr.push({ key: 'c', name: 'Godhuda' });
-  reference.update(arr);
+      assert.deepEqual(target.toValues(), objectValues(obj));
+    });
 
-  sync(target, artifacts);
+    test('it handles null delegates', assert => {
+      // Passing null will return an empty iterator
+      let target = new IterableWrapper(null);
 
-  assert.deepEqual(target.toValues(), arr);
+      assert.deepEqual(target.toValues(), []);
+    });
+  });
 
-  arr = arr.slice();
-  arr.shift();
-  reference.update(arr);
+  module('keys', () => {
+    test('@identity works', assert => {
+      let arr = [
+        { key: 'a', name: 'Yehuda' },
+        { key: 'b', name: 'Godfrey' },
+      ];
+      let target = new IterableWrapper(arr);
 
-  sync(target, artifacts);
+      assert.deepEqual(target.toKeys(), arr);
+    });
 
-  assert.deepEqual(target.toValues(), arr);
-});
+    test('@identity works with multiple values that are the same', assert => {
+      let yehuda = { key: 'a', name: 'Yehuda' };
+      let godfrey = { key: 'b', name: 'Godfrey' };
+      let arr = [yehuda, godfrey, godfrey];
 
-QUnit.test('When re-iterated via swap #1, the original references are updated', assert => {
-  let arr = getInitialArray(8, 1, 'i-');
-  let { target, reference, artifacts } = initialize(arr);
-  assert.deepEqual(target.toValues(), arr);
+      let target = new IterableWrapper(arr);
 
-  let a = arr[1];
-  let b = arr[7];
-  arr[7] = a;
-  arr[1] = b;
+      let keys1 = target.toKeys();
 
-  reference.update(arr);
+      assert.equal(keys1.length, 3);
+      assert.equal(keys1[0], yehuda);
+      assert.equal(keys1[1], godfrey);
 
-  target.cleanHistory();
-  sync(target, artifacts);
+      arr.pop();
+      arr.unshift(godfrey);
 
-  assert.equal(
-    target.serializeHistory(),
-    'retain:1,move:8,retain:3,retain:4,retain:5,retain:6,retain:7,move:2'
-  );
-  assert.equal(target.historyStats.move, 2, 'moved nodes count');
```

Comment:
this one ^

Reply:
I added some new instrumentation to bring back the coverage from these tests. Originally I thought the tests were only testing the test harness, but in reality they were in fact testing the actual behavior (since the `IteratorSynchronizer` was a separate class), and it would have reduced coverage to remove them entirely.

I also caught some issues thanks to the tests, my understanding of the algorithm was flawed because I wasn't taking into account the fact that the list was changing _as it was synchronizing_. I updated the algorithm and compared it using counts, and it appears to be as optimal as the previous one.

We should try to test the outcome of this is Krausest when we land it.